File: 2 kyu/Cut the cake/solution.rb
def cut(cake)
  map = cake.split("\n")
  cake_h = map.size
  cake_w = map.first.size
  n = cake.count('o')
  return [] unless (cake_h * cake_w % n).zero?
  piece_square = cake_h*cake_w/n
  widths = (1..piece_square).select{|w| (piece_square % w).zero?}.reverse
  result = []
  pieces = Array.new(cake_h){Array.new(cake_w){0}}
  
  cut = ->(k){
    return true if k == n  
    start_row = pieces.index{|row| row.include?(0)}
    start_col = pieces[start_row].index(0)
    widths.each{|w|
      h = piece_square/w
      rows = start_row...start_row+h
      cols = start_col...start_col+w
      if cols.end <= cake_w && rows.end <= cake_h &&
        pieces[rows].all?{|row| row[cols].all?{|x| x.zero?}} &&
        map[rows].sum{|row| row[cols].count('o')} == 1
        
        rows.each{|r| cols.each{|c| pieces[r][c] = k + 1}}
        result << [rows,cols]
        return true if cut.(k+1)
        result.pop
        rows.each{|r| cols.each{|c| pieces[r][c] = 0}}
      end  
    }
    false
  }
  cut.(0) 
  result.map{|(rows,cols)| rows.map{|r| map[r][cols]}.join("\n")}
end
_____________________________________________________
def cut cake
    k = cake.count('o')
    cake = cake.split("\n")
    n, m = cake.size, cake[0].size
    area = n * m
    area, r = area.divmod(k)
    return [] unless r.zero?

    dims = (1..[Math.sqrt(area).floor, n].min).each_with_object([]) do |i, a|
        next if area % i > 0
        j = area / i
        next if j > m
        a << i
        a << j unless i == j
    end
    dims.sort!

    accum = []
    recf = -> cake do
        res = y = nil
        if (x = cake.find_index{|r| y = /\S/ =~ r})
            dims.each_with_index do |k, i|
                l = dims[~i]
                if x+k <= n and y+l <= m
                    slice = cake[x...x+k].map{|r| r[y...y+l]}.join("\n")
                    if not slice.include?(' ') and slice.count('o') == 1
                        accum << slice
                        re = /(?<=.{#{y}})\S{#{l}}/
                        sliced = cake.map.with_index{|r,i| x <= i && i < x+k ? r.sub(re, ' ' * l) : r}
                        break if (res = recf.(sliced))
                        accum.pop
                    end
                end
            end
            res
        else
            true
        end
    end
    
    recf.(cake) ? accum : []
end
_____________________________________________________
def possible_rectangles(data, size)
  rectangles = []
  data[0].size.downto(1) do |width|
    data.size.downto(1) do |height|
      rectangles << [width, height] if width * height == size
    end
  end
  rectangles
end

def empty_position?(arr, y, x)
  arr.all? { |a| a[y][x].nil? }
end

def get_empty_position(arr)
  width = arr[0][0].size
  height = arr[0].size
  (0...height).each do |y|
    (0...width).each do |x|
      return [y, x] if arr.all? { |b| b[y][x].nil? }
    end
  end
  nil
end

def get_layer(arr, position, size, data)
  width, height = size
  y, x = position
  return nil if y + height > data.size || x + width > data[y].size

  layer = Array.new(data.size) { Array.new(data[0].size) }
  raisins = 0
  data.each_with_index do |row, yy|
    row.each_with_index do |column, xx|
      next unless
        yy.between?(y, y + height - 1) && xx.between?(x, x + width - 1)
      return nil unless empty_position?(arr, yy, xx)

      raisins += 1 if column == 'o'
      layer[yy][xx] = column
    end
  end
  return nil if raisins != 1

  layer
end

def search(raisins, arr, rectangles, data)
  return arr if raisins.zero?

  position = get_empty_position(arr)
  rectangles.each do |size|
    layer = get_layer(arr, position, size, data)
    next if layer.nil?

    arr << layer
    res = search(raisins - 1, arr, rectangles, data)
    return res unless res.nil?

    arr.pop
  end
  nil
end

def extract(board)
  board[1..-1].map do |layer|
    layer.map(&:join).reject(&:empty?).join("\n")
  end
end

def cut(cake)
  raisins = cake.count('o')
  data = cake.each_line.map { |line| line.strip.split('') }
  cells = data.size * data[0].size
  return [] unless (cells % raisins).zero?

  result = search(
    raisins,
    [Array.new(data.size) { Array.new(data[0].size) }],
    possible_rectangles(data, cells / raisins),
    data
  )
  result ? extract(result) : []
end

File: 2 kyu/Cut the cake/solution.hs
module Kata.Cut.Cake (cut) where

type CakeCut = ((Int, Int), (Int, Int))

cut :: String -> [String]
cut cake = reverse $ map (unlines . cutToPieces) $ 
    cutDaCake [[]] raisins
  where
    raisins = length $ filter (=='o') cake
    height = length actCake
    width = length (head actCake)
    cutSize = height * width `div` raisins
    actCake = lines cake
    cutToPieces :: CakeCut -> [String]
    cutToPieces ((y, x), (y', x'))
      | y > y'    = []
      | otherwise = drop x (take (x' + 1) (actCake !! y)) : cutToPieces ((y + 1, x), (y', x'))
    pieces :: (Int, Int) -> [CakeCut] -> [CakeCut]
    pieces (y,x) taken = [((y, x), (y', x')) | y' <- [y..height-1], x' <- [x..width-1],
        (y' - y + 1) * (x' - x + 1) == cutSize,
        containsRaisins y x y' x' [] == 1,
        notTaken y x x' taken]
    containsRaisins :: Int -> Int -> Int -> Int -> String -> Int
    containsRaisins y x y' x' some
      | null some && y > y' = 0
      | null some           = containsRaisins (y + 1) x y' x' (drop x (take (x' + 1) (actCake !! y)))
      | otherwise           = (if head some == 'o' then 1 else 0) + containsRaisins y x y' x' (tail some)
    notTaken :: Int -> Int -> Int -> [CakeCut] -> Bool
    notTaken y x x' taken = not (any (\((_, tx), (_, tx')) -> x <= tx && tx <= x' || tx <= x && x <= tx')
        (filter ((>=y) . fst . snd) taken))
    corners = [(y, x) | y <- [0..height-1], x <- [0..width-1]]
    cutDaCake :: [[CakeCut]] -> Int -> [CakeCut]
    cutDaCake soFarTaken m
      | m == 0    = if null soFarTaken then [] else head soFarTaken
      | otherwise = cutDaCake (findNewPiece soFarTaken) (m - 1)
      where
        findNewPiece :: [[CakeCut]] -> [[CakeCut]]
        findNewPiece list = concatMap (\taken -> [piece:taken | piece <- pieces (next taken) taken]) list
          where
            next taken = head $ filter (\(y, x) -> notTaken y x x taken) corners
            
_____________________________________________________
module Kata.Cut.Cake (cut) where

import qualified Data.Set as S (Set,fromList,elemAt,member,difference)

cut :: String -> [String]
cut cake' = if null sols then [] else head $ sols
  where (nx,ny) = (length.words $ cake',length.head.words $ cake')
        cake = map (flip divMod ny.fst).filter ((=='o').snd).zip [0..].filter (/='\n') $ cake'
        n = length cake
        a = nx*ny `div` n
        rs = recs 1
        recs k | k^2>=a = if k^2==a then [(k,k)] else []
               | True = let (k',r) = a `divMod` k in if r==0 then (k,k'):(k',k):recs (k+1) else recs (k+1) 
        f s k l | k==n = map reverse [l]
                | True = concat [f (S.difference s (S.fromList [(x+i,y+j) | i<-[0..dx-1], j<-[0..dy-1]])) (k+1) (((x,y),(dx,dy)):l)| (dx,dy)<-rs']
          where (x,y) = S.elemAt 0 s
                rs' = filter (fit s (x,y)) rs
        fit s (x,y) (dx,dy) = c1 && c2
          where c1 = and.map (flip S.member s) $ [(x+i,y+j) | i<-[0..dx-1], j<-[0..dy-1]]
                c2 = (==1).length.filter (\(a,b) -> x<=a && a<x+dx && y<=b && b<y+dy) $ cake
        sols = map (map (\((x,y),(dx,dy)) -> concat [[cake' !! (i*(ny+1)+j) | j<-[y..y+dy-1]] ++ "\n" | i<-[x..x+dx-1]])) $
               f (S.fromList [(i,j) | i<-[0..nx-1], j<-[0..ny-1]]) 0 []

_____________________________________________________
{-# LANGUAGE TupleSections #-}
module Kata.Cut.Cake (cut) where

import qualified Data.Array as Array
import qualified Data.Set as Set
import qualified Data.List as List
import qualified Data.Maybe as Maybe
import qualified Data.List.Split as Split
import Data.Array((//), (!))

cut :: String -> [String]
cut cake = if areaMod == 0
         then case place Set.empty [] (Array.indices board) pieceCount of
                (Just pieces) -> map (showBoard . cutBoard board) pieces
                Nothing -> []
         else []
  where board = parseBoard cake
        raisins = Set.fromList $ map fst $
          filter ((==CellRaisin) . snd) $ Array.assocs board
        area =
          let (_, (lastRow, lastCol)) = Array.bounds board
           in (lastRow+1)*(lastCol+1)
        pieceCount = Set.size raisins
        (pieceArea, areaMod) = area `divMod` pieceCount
        sizes = map (\v -> (v, pieceArea `div` v)) $ divisors pieceArea
        place _ pieces _ 0 = Just $ reverse pieces
        place _ _ [] _ = Nothing
        place seen pieces (p:ps) cnt =
          let posRect = makeRectangle p (1,1)
           in if noIntersections pieces posRect
             then let cands = Maybe.mapMaybe ((\rect ->
                       if canPlace pieces rect
                          then (rect,) <$> singleRaisin seen rect
                          else Nothing) . makeRectangle p) sizes
                   in case Maybe.mapMaybe (\(rect, raisin) -> place (Set.insert raisin seen) (rect:pieces) ps (cnt-1)) cands of
                        (v:_) -> Just v
                        [] -> Nothing
             else place seen pieces ps cnt
        singleRaisin seen rect =
          let cutRaisins = filter (rectanglesIntersect rect . flip makeRectangle (1, 1)) $ Set.toList raisins
           in if (length cutRaisins == 1) && not (Set.member (head cutRaisins) seen)
                 then Just $ head cutRaisins
                 else Nothing
        canPlace pieces rect =
          let (_, (lastRow, lastCol)) = Array.bounds board
              (rectBottom, rectRight) = rectangleBottomRight rect
           in rectBottom <= lastRow && rectRight <= lastCol && noIntersections pieces rect
        noIntersections targets rect =
          all (not . rectanglesIntersect rect) targets

type Vect = (Int, Int)

data Rectangle = Rectangle Vect Vect

data Cell = CellEmpty | CellRaisin
  deriving Eq

makeRectangle posTopLeft@(row, col) (height, width) =
  Rectangle posTopLeft (row+height-1, col+width-1)
rectangleBottomRight (Rectangle _ posBottomRight) = posBottomRight

parseBoard :: (Integral a, Array.Ix a) => String -> Array.Array (a, a) Cell
parseBoard board =
  let ls = lines board
      height = fromIntegral $ length ls
      width = maybe 0 (fromIntegral . length . fst) $ List.uncons ls
   in if height > 0 && width > 0
         then Array.listArray ((0, 0), (height-1, width-1)) $
           map (\v -> if v == 'o' then CellRaisin else CellEmpty) $ concat ls
         else Array.listArray ((0, 0), (0, 0)) [CellEmpty]

showBoard :: (Integral a) => Array.Array (a, a) Cell -> String
showBoard board =
  let (_, (_, lastCol)) = Array.bounds board
   in List.intercalate "\n" (Split.chunksOf (fromIntegral $ lastCol+1) $
     map showCell $ Array.elems board) ++ ['\n']
       where showCell CellEmpty = '.'
             showCell CellRaisin = 'o'

cutBoard :: Array.Array (Int, Int) Cell -> Rectangle -> Array.Array (Int, Int) Cell
cutBoard board (Rectangle (rb, cb) (re, ce)) =
  let res = Array.listArray ((0, 0), (re-rb, ce-cb)) $ repeat CellEmpty
   in res // map (\idx@(r, c) -> (idx, board!(rb + r, cb + c))) (Array.indices res)

rectanglesIntersect :: Rectangle -> Rectangle -> Bool
rectanglesIntersect (Rectangle (rb1, cb1) (re1, ce1)) (Rectangle (rb2, cb2) (re2, ce2)) =
  intersect (rb1, re1) (rb2, re2) && intersect (cb1, ce1) (cb2, ce2)
    where intersect (b1, e1) (b2, e2) =
            let s1 = signum (e2-b1)
                s2 = signum (b2-e1)
             in s1 /= s2 || s1 == 0

divisors val = List.sort $ concatMap (\v -> List.nub [v, val `div` v]) $
  filter ((==0) . (val`mod`)) $ takeWhile ((<=val) . (^2)) [1..]
_____________________________________________________
module Kata.Cut.Cake (cut) where
import Data.List
cut :: String -> [String]
cut cake
 | null cake||r==0||mod (m*n) r/=0||null solution = [] -- special cases where no solution exist
 | otherwise = head solution -- take the first solution that you find
 where x = lines cake -- convert the cake into a list of lines
       m = length (head x) -- width of the cake
       n = length x -- height of the cake
       r = length (filter (=='o') cake) -- number of raisins
       ar = div (m*n) r -- area of each piece
       all = [(i,j)|j<-[0..(n-1)],i<-[0..(m-1)]] -- all coordinates of x
       solution = [s|s<-go all r, length s==r] -- look for a list of r pieces that can be cut off
       go :: [(Int,Int)] -> Int -> [[String]]
       go rest s -- return all lists of pieces that can be cut off
        | null next = [[]] -- no piece can be cut off anymore
        | otherwise =  [(piece (div ar d) d):l|d<-next, l<-go (rest\\[(a+i,b+j)|i<-[0..(div ar d -1)],j<-[0..(d-1)]]) (r-1)]
        where (a,b) = head rest -- position of the next piece
              mw = head [w|w<-[1..], (a+w,b) `notElem` rest] -- maximal width of the next piece
              mh = head [h|h<-[1..], (a,b+h) `notElem` rest] -- maximal width of the next piece
              divisors = [d|d<-[1..mh], mod ar d==0, div ar d<=mw] -- possible heights of the next piece
              piece w h = unlines [take w (drop a (x!!j))|j<-[b..(b+h-1)]] -- returns a piece with given width and height
              next = [d|d<-divisors, length (filter (=='o') (piece (div ar d) d))==1] -- possible heights such that the next piece only contains one raisin

